crates/wasmparser/src/validator/operators.rs

```diff
@@ -1409,160 +1550,175 @@ impl OperatorValidator {
             }
             Operator::RefNull { ty } => {
                 self.check_reference_types_enabled()?;
-                match ty {
-                    ValType::FuncRef | ValType::ExternRef => {}
-                    _ => {
-                        return Err(OperatorValidatorError::new(
-                            "invalid reference type in ref.null",
-                        ))
-                    }
-                }
-                self.push_operand(ty)?;
+                self.push_operand(
+                    ValType::Ref(RefType {
+                        nullable: true,
+                        heap_type: ty,
+                    }),
+                    resources,
+                )?;
             }
             Operator::RefIsNull => {
                 self.check_reference_types_enabled()?;
-                match self.pop_operand(None)? {
-                    None | Some(ValType::FuncRef) | Some(ValType::ExternRef) => {}
-                    _ => {
-                        return Err(OperatorValidatorError::new(
-                            "type mismatch: invalid reference type in ref.is_null",
-                        ))
-                    }
-                }
-                self.push_operand(ValType::I32)?;
+                self.pop_ref(resources)?;
+                self.push_operand(ValType::I32, resources)?;
             }
             Operator::RefFunc { function_index } => {
                 self.check_reference_types_enabled()?;
-                if resources.type_of_function(function_index).is_none() {
-                    return Err(OperatorValidatorError::new(format!(
-                        "unknown function {}: function index out of bounds",
-                        function_index,
-                    )));
-                }
+                let type_index =
+                    if let Some(type_index) = resources.type_index_of_function(function_index) {
+                        type_index
+                    } else {
+                        return Err(OperatorValidatorError::new(format!(
+                            "unknown function {}: function index out of bounds",
+                            function_index,
+                        )));
+                    };
                 if !resources.is_function_referenced(function_index) {
                     return Err(OperatorValidatorError::new("undeclared function reference"));
                 }
-                self.push_operand(ValType::FuncRef)?;
+                if self.features.function_references {
```

Comment:
Given that the type-checking here is purely internal to this crate, I think it should be ok to remove this `if` and always push the non-nullable type, right?

Reply:
`push_operand` will validate that the type is valid with respect to the enabled features, which will fail when the feature isn't enabled and the reference is non-nullable.   This could be changed, it just matches the existing pattern I noticed

Reply:
Ah ok that makes sense. In general we try to avoid control flow based on enabled features though. Ideally unconditionally enabling a feature in `wasmparser` would simply be "remove all the `if !enabled { return err }` branches" as otherwise we run the possibility of having divergent behavior in the validator whether or not the feature is enabled or not. I don't think that this individual block would result in any issues, but if possible I think it would be good to allow `RefType` in the validator and only validate that user-provided types (e.g. function types, block types) don't contain `RefType` if the function-references proposal is disabled.